wfe/wfe.go

```diff
@@ -1852,27 +1868,37 @@ func (wfe *WebFrontEndImpl) updateChallenge(
 		return
 	}
 
-	var chalResp struct {
-		KeyAuthorization *string
-	}
-	err := json.Unmarshal(postData.body, &chalResp)
-	if err != nil {
+	// In strict mode we reject any challenge POST with a body other than `{}`.
+	// This matches RFC 8555 Section 7.5.1.
+	if wfe.strict && !bytes.Equal(postData.body, []byte("{}")) {
 		wfe.sendError(
-			acme.MalformedProblem("Error unmarshaling body JSON"), response)
+			acme.MalformedProblem(`challenge initiation POST JWS body was not "{}"`), response)
 		return
-	}
+	} else {
+		// When not in strict mode we still want to be strict about the legacy key
+		// authorization field not being present in the POST JSON.
+		var chalResp struct {
+			KeyAuthorization *string
+		}
+		err := json.Unmarshal(postData.body, &chalResp)
+		if err != nil {
+			wfe.sendError(
+				acme.MalformedProblem("Error unmarshaling body JSON"), response)
+			return
+		}
 
-	// Historically challenges were updated by POSTing a KeyAuthorization. This is
-	// unnecessary, the server can calculate this itself. We could ignore this if
-	// sent (and that's what Boulder will do) but for Pebble we'd like to offer
-	// a way to be more aggressive about pushing clients implementations in the
-	// right direction, so we treat this as a malformed request.
-	if chalResp.KeyAuthorization != nil {
-		wfe.sendError(
-			acme.MalformedProblem(
-				"Challenge response body contained legacy KeyAuthorization field, "+
-					"POST body should be `{}`"), response)
-		return
+		// Historically challenges were updated by POSTing a KeyAuthorization. This is
```

Comment:
It looks like we already reject challenge POSTs with "KeyAuthorization", even in non-strict mode. So we don't need to guard the new code with an `if strict`. We can just make it more aggressive if we want. Alternately we can consider the existing KeyAuthorization rejection code as good enough, when combined with the new `resource` rejection, since there were no other fields in common use for challenge posts.

Reply:
I think you might have read the diff wrong. The key auth reject is in the `else` of a check for strict. E.g. it only happens when strict is disabled. I could have left it outside of the `else` entirely but it felt silly to check for a key auth field in the case where we've already verified the body is `{}` for strict.

Reply:
We do need to guard the new code with `-strict` because otherwise integration tests will fail.

Reply:
Oh yeah, I get it now. I was thinking of `resource` as already handled by the auth code, but it also comes into play here, and is still send by `acme`.